src/spesh/optimize.c

```diff
@@ -1065,60 +951,182 @@ static void optimize_istrue_isfalse(MVMThreadContext *tc, MVMSpeshGraph *g, MVMS
-                }
-                /* We can just unbox the int and pretend it's a bool. */
-                ins->info = MVM_op_get_op(MVM_OP_unbox_i);
-                if (negated_op)
-                    ins->operands[0] = temp;
-                /* And then we might be able to optimize this even further. */
-                optimize_repr_op(tc, g, bb, ins, 1);
-                break;
-            case MVM_BOOL_MODE_NOT_TYPE_OBJECT:
-                /* This is the same as isconcrete. */
-                ins->info = MVM_op_get_op(MVM_OP_isconcrete);
-                if (negated_op)
-                    ins->operands[0] = temp;
-                /* And now defer another bit of optimization */
-                optimize_isconcrete(tc, g, ins);
+        switch (mode) {
+        case MVM_BOOL_MODE_ITER:
+            if (!guaranteed_concrete)
                 break;
-            /* TODO implement MODE_UNBOX_NUM and the string ones */
-            default:
+            if (input_facts->flags & MVM_SPESH_FACT_ARRAY_ITER) {
+                ins->info = MVM_op_get_op(MVM_OP_sp_boolify_iter_arr);
+            } else if (input_facts->flags & MVM_SPESH_FACT_HASH_ITER) {
+                ins->info = MVM_op_get_op(MVM_OP_sp_boolify_iter_hash);
+            } else {
+                ins->info = MVM_op_get_op(MVM_OP_sp_boolify_iter);
+            }
+            break;
+        case MVM_BOOL_MODE_UNBOX_INT:
+            if (!guaranteed_concrete)
+                return;
+                /* We can just unbox the int and pretend it's a bool. */
+            ins->info = MVM_op_get_op(MVM_OP_unbox_i);
+            /* And then we might be able to optimize this even further. */
+            optimize_unbox(tc, g, bb, ins);
+            break;
+        case MVM_BOOL_MODE_BIGINT:
+            if (!guaranteed_concrete)
                 return;
+            ins->info = MVM_op_get_op(MVM_OP_bool_I);
+            break;
+        case MVM_BOOL_MODE_HAS_ELEMS:
+            if (!guaranteed_concrete)
+                return;
+            ins->info = MVM_op_get_op(MVM_OP_elems);
+            optimize_repr_op(tc, g, bb, ins, 1);
+            break;
+        case MVM_BOOL_MODE_NOT_TYPE_OBJECT:
+            ins->info = MVM_op_get_op(MVM_OP_isconcrete);
+            /* And now defer another bit of optimization */
+//            optimize_isconcrete(tc, g, ins);
+            break;
+            /* We need to change the register type for our result for this,
+             * means we need to insert a temporarary and a coerce:
+        case MVM_BOOL_MODE_UNBOX_NUM:
+             op_info = MVM_op_get_op(MVM_OP_unbox_i);
+             break;
+            */
+        default:
+            return;
         }
-        /* Now we can take care of the negation. */
+//        fprintf(stderr, "Known type optimized to %s\n", ins->info->name);
+        /* Now we can take care of the negation. - NB I'm not entirely sure why
+         * this would need it's own register though! */
```

Comment:
Because otherwise we end up with multiple assigns to the same version of the register and violate SSA form.

Reply:
Yeah, I figured that out, I'll add a comment to that effect.
We could've had a 'new version' rather than an entirely new register, though, that would have had the same effect.